### PR TITLE
Support Blogger categories

### PR DESCRIPTION
**Pull request by ScottIsAFool**

### CODE REVIEW

**Comment by martinwoodward** [comment]
Looks good - what's left?

**Comment by ScottIsAFool** [comment]
Not much, just
image

**Comment by kathweaver** [comment]
So far, you haven't broken Typepad, and I am able to load categories from Blogger.

Yes, I've opened some test blog accounts

**Comment by willduff** [comment]
I tested the categories API call that Scott added against a Blogger blog hosted on the blogspot.se domain and another one hosted on a custom domain. Both blogs responded with a complete list of a categories in the expected JSON format - nice find Scott!

I posted a **draft** to my own Blogger blog that contained a new category in it. I noticed that if I then make a request to the categories API that Scott added, the new category from my draft isn't there. But it shows up as available in the Blogger.com dashboard as a known label for any blog post. This makes sense because the categories API that Scott added is a public URL, so revealing users' unpublished categories would be a privacy concern. But this means that saving a local draft and then reopening it later might lose categories. I confirmed I do loose categories this way, but that actually might be due to this second issue I'm hitting.

After closing Writer and then re-opening it, the categories list and the autocomplete in the UI isn't working anymore. I re-installed and I can reliably reproduce this - categories work great on first run but not thereafter. The first time I used categories, I got this nice drop down like this:
image

Now after restarting Writer I just see a refresh button, but when I click on it I don't see any HTTP requests in Fiddler:
image

I'll try and debug through this tomorrow.

**Comment by willduff** [comment]
I built the installer myself from source and now it's working fine...still looking.

**Comment by kathweaver** [comment]
Just downloaded and installed from the artifacts here, with no issues.

**Comment by kathweaver** [comment]
Posted a post to blogger (blogger domain), and added a category.  Worked as expected.

**Comment by kathweaver** [comment]
Having an issue with posting to Typepad.


[Open Live Writer.txt](https://github.com/OpenLiveWriter/OpenLiveWriter/files/91306/Open.Live.Writer.txt)

**Comment by willduff** [comment]
@kathweaver What if you quit out of Open Live Writer and then launch it again and do another post. Does adding categories show the drop down?
image

**Comment by willduff** [comment]
@kathweaver Your TypePad issue seems to be a networking problem. For some reason TypePad.com is resolving as 192.168.1.254 for you, but that IP address range is private and will generally resolve to your home network.

From your log:
```
Network Connection Error - Error attempting to connect to blog at: http://www.typepad.com/services/xmlrpc
Unable to connect to the remote server
No connection could be made because the target machine actively refused it 192.168.1.254:8080
```

**Comment by willduff** [comment]
Ah, figured it out! Scott's branch version is out of date, so I'm actually getting upgraded from 0.5.1.3 to 0.5.1.4 when I quit out of Writer. That won't matter when I merge this into master...which I'm going to do now!

File: src/managed/OpenLiveWriter.BlogClient/Clients/GoogleBloggerv3Client.cs [hunk 1]
```diff
@@ -88,7 +88,7 @@ private static BlogPost ConvertToBlogPost(Post post)
                 Permalink = post.Url,
                 Contents = post.Content,
                 DatePublished = post.Published.Value,
-                Keywords = string.Join(new string(LabelDelimiter,1), post.Labels ?? new List<string>())
+                Categories = post.Labels.Select(x => new BlogPostCategory(x)).ToArray()
```
Comment (willduff) [inline src/managed/OpenLiveWriter.BlogClient/Clients/GoogleBloggerv3Client.cs]: This needs to be `post.Labels?.Select(x => new BlogPostCategory(x)).ToArray()`. post.Labels can be null, so I'm getting a NullReferenceException on some posts when I go to **File** > **Open recent post** > **My Test Blog**.

Comment (willduff) [inline src/managed/OpenLiveWriter.BlogClient/Clients/GoogleBloggerv3Client.cs]: @kathweaver If its not captured already, opening posts from the remote blog server would be good to add to the test plan. You can do this via  **File** > **Open recent post** > **Your Blog Name**. Then you can browse through and select any post, and Writer will download the post from the blog server and open it as a local draft.

**Comment by willduff** [comment]
Found one more issue worth fixing before we take this.

**Comment by kathweaver** [comment]
@willduff  Yes, that test plan is there:  In my latest branch, I moved that under \testplan\testFile and in observeOpenDialogBox.  It might be worthwhile to split things out further.

**Comment by kathweaver** [comment]
Never did figure out the TypePad issue.  I rebooted, deleted everything related to Open File Writer and was able to post again.  Didn't go back and pull and test this again as it got late.

**Comment by kathweaver** [comment]
I'm running to the same problem you are -- when I run the setup, I can edit the categories fine, but if I get in and out, I go back to the previous release -- I think I need to unselect auto updates.

**Comment by ScottIsAFool** [comment]
I've just fixed up a few potential ArgumentNullExceptions and updated the PR

**Comment by kathweaver** [comment]
Ran a few tests, and so far you haven't broken anything.

**Comment by willduff** [comment]
Looks good to me! Will merge this in shortly.

**Comment by Indy27** [comment]
I wonder is it possible to enlarge the drop-down box (height, not width), or make the box into two columns for easy scrolling?

**Pull request merged**